### ISSUE TITLE
Stylelint CLI does NOT use custom processor when needed.

### ISSUE THREAD

**Issue by surgeboris** [stylelint/stylelint#4223]
> Clearly describe the bug

Stylelint CLI does NOT use custom processor when I'm passing file contents via stdin. However when I'm passing filename as a stylelint CLI argument - it uses custom processor correctly. I prepared [a repo with minimal reproduction](https://github.com/surgeboris/styleint-bug-repro).

> What stylelint configuration is needed to reproduce the bug?

```json
{
  "processors": [
    ["stylelint-processor-styled-components"]
  ],
  "extends": [
    "stylelint-config-recommended",
    "stylelint-config-styled-components"
  ]
}
```

> Which version of stylelint are you using?

 `10.1.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

```bash
    surgeboris$ cat ComponentWithError.tsx | stylelint --config ./.stylelintrc
    
    <input css 1>
     3:77  ✖  Unexpected token (3:77)

    surgeboris$ stylelint --config ./.stylelintrc ComponentWithError.tsx
    
    ComponentWithError.tsx
     6:5  ✖  Unexpected unknown property "wuidth"   property-no-unknown 
```
Problem is also reproduced without explicit `--config` flags.

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

I'm using styled-components with Typescript.

> What did you expect to happen?

I expect stylelint CLI to use custom processor regardless of whether the file passed via stdin or via filename.

> What actually happened (e.g. what warnings or errors did you get)?

IIUC when file contents are passed via stdin stylelint uses default `CSS-in-JS` syntax processor which easily deals with javascript and jsx (and hence with some subset of typescript and tsx), but throws error if more obscure typescript features are used (e.g. property-access on a type, non-null assertions). When file contents are passed as a filename argument - proper custom syntax processor is used (stylelint-processor-styled-components).

**Comment by hudochenkov** [comment]
There is no way to find out what syntax if you just pass stdin. Use `--syntax css-in-js` parameter:

> cat ComponentWithError.tsx | stylelint --config ./.stylelintrc --syntax css-in-js

**Comment by surgeboris** [comment]
> There is no way to find out what syntax if you just pass stdin. Use --syntax css-in-js parameter

@hudochenkov But I'm using "processors" property in the configuration file, and passing config argument explicitly. Isn't that enough to find out what syntax I'm using?

Anyway, I've tried `--syntax css-in-js` and it does not help, since built-in css-in-js parser does not handle more obscure typescript features (e.g. property-access on a type, non-null assertions) very well.

**Comment by surgeboris** [comment]
Passing additional `--stdin-filename ComponentWithError.tsx` parameter when passing file contents via stdin actually solves the problem.

Which makes me wonder why is it necessary to know the filename to use the custom processor that is in explicitly passed config...

I'm also struggling to understand the difference between `syntax` and `processor` concepts: IIUC `syntax` is some kind of a built-in out-of-the-box processor. So we need the filename to know the `syntax`, but then it's anyway overriden by `processor`? It would be cool to get some explanation why it works like that (or to be pointed on where this explanation can be found).

Anyway, `--stdin-filename` solves my problem (it was about text editor integration plugin, so I always know when which filename to pass as `--stdin-filename`) and this issue may be closed.
